### ISSUE TITLE
noalias on a Box doesn't seem to have an impact when compiling without `-C panic=abort`

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/rust#95323]
I tried this code:

```rust
struct Foo {
    x: i32
}
#[inline(never)]
fn function2() { println!("foo"); }

#[inline(never)]
fn function1() -> Box<Foo> { Box::new(Foo { x: 4}) }

pub fn foo() {
    let foo: Box<_> = function1();
    let value1 = foo.x;
    function2();
    let value2 = foo.x;
    assert_eq!(value1, value2);
}
```

I expect the assert_eq! to be compiled away. However, this only happens with `-C panic=abort`.

I'm using rustc 1.61.0-nightly (c84f39e6c 2022-03-20)

**Comment by RalfJung** [comment]
Possibly related: https://github.com/rust-lang/rust/issues/46515

**Comment by ldm0** [comment]
Cannot reproduce with nightly rust: https://godbolt.org/z/M5vvsbGnK

**Comment by RalfJung** [comment]
Cc @Nilstrieb

**Comment by nikic** [comment]
This always worked with `-C opt-level=3` but failed (and still fails) with `-C opt-level=2`.

**Comment by Nilstrieb** [comment]
I think LLVM inlined the function anyways and that's why it's not reproducible anymore (even with `-Zbox-noalias=no`). Interestingly, if the function is made truly opaque my passing a function pointer instead, the assert is _not_ optimized away with `-Zbox-noalias=no`, but it is optimized away with `-Zbox-noalias=yes`. So this indeed seems like a case where `noalias` on box improves codegen.

**Comment by nikic** [comment]
@Nilstrieb Yes, LLVM infers that the return value is noalias in this case (noinline does not affect IPO inference passes).
